File: client-android/serenada-core/src/main/java/app/serenada/core/CallStats.kt
package app.serenada.core

import app.serenada.core.call.RealtimeCallStats

/**
 * Aggregated call statistics exposed to SDK consumers.
 * Populated from WebRTC getStats() periodically during an active call.
 */
data class CallStats(
    /** Aggregate bitrate in kbps. */
    val bitrate: Double? = null,
    /** Packet loss percentage (0-100). */
    val packetLoss: Double? = null,
    /** Network jitter in milliseconds. */
    val jitter: Double? = null,
    /** Active audio/video codec name. */
    val codec: String? = null,
    /** Active ICE candidate pair description. */
    val iceCandidatePair: String? = null,
    /** Network round-trip time in milliseconds. */
    val roundTripTime: Double? = null,
    /** Audio receive bitrate in kbps. */
    val audioRxKbps: Double? = null,
    /** Audio transmit bitrate in kbps. */
    val audioTxKbps: Double? = null,
    /** Video receive bitrate in kbps. */
    val videoRxKbps: Double? = null,
    /** Video transmit bitrate in kbps. */
    val videoTxKbps: Double? = null,
    /** Current video frames per second. */
    val videoFps: Double? = null,
    /** Current video resolution (e.g. "1280x720"). */
    val videoResolution: String? = null,
    /** Real-time per-frame statistics. */
    val realtimeStats: RealtimeCallStats? = null,
    /** Timestamp of last stats update (epoch millis). */
    val updatedAtMs: Long = 0L,
)

File: client-ios/SerenadaCore/Sources/Models/CallStats.swift
import Foundation

/// Aggregated WebRTC call statistics.
public struct CallStats: Equatable {
    /// Available outgoing bitrate in kbps.
    public var bitrate: Double?
    /// Packet loss percentage (video or audio).
    public var packetLoss: Double?
    /// Audio jitter in milliseconds.
    public var jitter: Double?
    /// Active audio/video codec name.
    public var codec: String?
    /// Active ICE candidate pair description.
    public var iceCandidatePair: String?
    /// Round-trip time in milliseconds.
    public var roundTripTime: Double?
    /// Audio receive bitrate in kbps.
    public var audioRxKbps: Double?
    /// Audio transmit bitrate in kbps.
    public var audioTxKbps: Double?
    /// Video receive bitrate in kbps.
    public var videoRxKbps: Double?
    /// Video transmit bitrate in kbps.
    public var videoTxKbps: Double?
    /// Incoming video frames per second.
    public var videoFps: Double?
    /// Incoming video resolution (e.g. "1280x720").
    public var videoResolution: String?
    /// Timestamp of the last stats update (epoch milliseconds).
    public var updatedAtMs: Int64 = 0

    public static let empty = CallStats()

    public init() {}

    public init(from realtimeStats: RealtimeCallStats) {
        self.bitrate = realtimeStats.availableOutgoingKbps
        self.packetLoss = realtimeStats.videoRxPacketLossPct ?? realtimeStats.audioRxPacketLossPct
        self.jitter = realtimeStats.audioJitterMs
        self.roundTripTime = realtimeStats.rttMs
        self.audioRxKbps = realtimeStats.audioRxKbps
        self.audioTxKbps = realtimeStats.audioTxKbps
        self.videoRxKbps = realtimeStats.videoRxKbps
        self.videoTxKbps = realtimeStats.videoTxKbps
        self.videoFps = realtimeStats.videoFps
        self.videoResolution = realtimeStats.videoResolution
        self.iceCandidatePair = realtimeStats.transportPath
        self.updatedAtMs = realtimeStats.updatedAtMs
    }
}
